### PR TITLE
Handle Network.Socket.connect: resource exhausted exceptions

### PR DESCRIPTION
**Pull request by rvl**
### Issue Number

Relates to #1708

### Overview

- Adds regression test
- Fixes unhandled exception resulting from parallel local snocket connections (see issue description for more details).

### CODE REVIEW

File: lib/shelley/src/Cardano/Wallet/Shelley/Network.hs [hunk 1]
```diff
@@ -804,6 +804,12 @@ handleIOException tr onResourceVanished e
     | isDoesNotExistError e =
         pure True
 
+    -- If the nonblocking UNIX domain socket connection cannot be completed
+    -- immediately (i.e. connect() returns EAGAIN), try again. This happens
+    -- because the node's listen queue is quite short.
+    | isTryAgainError e =
+        pure True
```
Comment (KtorZ) [inline lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:811]: Interesting :thinking:

File: lib/shelley/test/unit/Cardano/Wallet/Shelley/NetworkSpec.hs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+{-------------------------------------------------------------------------------
+                                      Spec
+-------------------------------------------------------------------------------}
+
+spec :: Spec
+spec = describe "NetworkLayer regression test #1708" $ do
+    it "Parallel local socket connections" $
+        withTestNode $ \np sock vData -> withLogging $ \(tr, _getLogs) ->
+            replicateConcurrently_ 5 $
```
Comment (KtorZ) [inline lib/shelley/test/unit/Cardano/Wallet/Shelley/NetworkSpec.hs:36]: Was this enough to observe the failure? (I assume so)

**Review by KtorZ — APPROVED** [review]
:+1: Nice!

**Comment by KtorZ** [comment]
Failure above is another instance of the unit test OOM issue, fixed in another PR.

File: lib/shelley/test/unit/Cardano/Wallet/Shelley/NetworkSpec.hs (hunk 2 above)
Comment (rvl) [reply lib/shelley/test/unit/Cardano/Wallet/Shelley/NetworkSpec.hs:36]: Yes - probably 3 concurrently would be enough, but why not 5.

**Pull request merged**